my/location/common.py

```diff
@@ -9,9 +10,14 @@
 
 
 # TODO: add timezone to this? can use timezonefinder in tz provider instead though
-class Location(NamedTuple):
+
+# converted from namedtuple to a dataclass so datasource field can be added optionally
+# if we want, can eventually be converted back to a namedtuple when all datasources are compliant
+@dataclass
+class Location:
     lat: float
     lon: float
     dt: datetime
     accuracy: Optional[float]
     elevation: Optional[float]
+    datasource: Optional[str] = None  # which module provided this, useful for debugging
```

Comment:
Yeah, something like a helper that adds a field to every NT/dataclass which could be used in `all.py`, but those tend to be readonly, so would likely need to do some sort of weird non-mypy-compliant magic with a proxy object which likely breaks cachew etc..., so not sure how automatic could be

For locations in particular I thought it'd be useful incase youre trying to exclude a faulty/wrong location geolocated from ip addresses for example

Reply:
> Would be kinda cool to have it for every data provider.. but haven't thought too much how to make it generic yet.

Apologies for the poor quality of this reply. I’m not feeling well and as much as I try I can’t understand the code that’s being worked on. I’m adding this in case it helps the project but I understand if it doesn’t fit.

When I was putting thought in to building something similar to HPI I was targeting a graph database so that each item would have (Item)->(Metadata) such as (jpg file)->(EXIF size), (jpg file)-(EXIF date), and so-on. That way each piece of metadata could have source (eg: “Josh’s iPhone”) and accuracy properties. In that scenario correcting the record was a matter of adding the new metadata node with a source of “User: Josh” and an accuracy of whatever they decided.
Because this type of correction is non-destructive there was just logic that pulled all sources of metadata from the API then displayed the user-defined source as the primary.

Reply:
The graph database idea makes a lot of sense as a solution to structuring this data, but HPI in general tends to denormalize databases into single dataclasses/namedtuples so that its able to be cached with `cachew` and just easier to manage as an `Iterator` instead of downstream functions (other functions, `hpi query`) having to know the specifics of every API

Still not sure what my plan is on a separate repo/how much to put there, but will see if I can get some basic file-based graphdb working